third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -126,10 +126,146 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
   return LLVM::AMD::llGetPid(loc, rewriter, moduleOp, axis);
 }
 
+static std::optional<Operation *> getReduxOpForDPP(triton::ReduceOp op) {
+  // In order to utilize the full power of DPP modifier, the redux operator
+  // should be a single arthmetic operator.
+  //
+  // Reference: third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TargetInfo.cpp
+  if (op.getNumOperands() != 1 || op.getNumResults() != 1)
+    return std::nullopt;
+  Block *block = &(*op.getCombineOp().begin());
+  Operation *yield = block->getTerminator();
+  Operation *reduceOp = yield->getOperand(0).getDefiningOp();
+  if (!reduceOp || reduceOp->getNumOperands() != 2 ||
+      reduceOp->getNumResults() != 1)
+    return std::nullopt;
+  if (reduceOp->getOperand(0) != block->getArgument(0) ||
+      reduceOp->getOperand(1) != block->getArgument(1))
+    return std::nullopt;
```

Comment:
What about defining a method on `tt.reduce` like `getSingleCombiner` so that we can extract his common part and unify with NVIDIA impl?

Reply:
This function assumes there's only 1 operator in the CombineOpso that we can extract it and try to fuse it with v_mov_dpp later. I'm not sure it's a commonly accepted assumption in the CombineOp in `triton::ReduceOp`. Like if we use shuffle & accumulate instead of the warp reduction here, it will just apply the whole CombineOp.

Reply:
That's fine--we are not changing the semantics of `tt.reduce` to only support such case. It's just adding a utility method for this common case which we need for both NVIDIA and AMD.

Reply:
I see. Yeah that definitely makes sense.